### PR TITLE
[Voice] Add build-it interpreter doc and fix dialog commands

### PR DESCRIPTION
**Pull request by GiviMAD**

### CODE REVIEW

**Comment by GiviMAD** [comment]
I would like to move voice documentation to its own page outside of multimedia. Is that ok?

File: configuration/multimedia.md [hunk 1]
```diff
@@ -220,7 +220,7 @@ There are two implementations available by default:
...
-| `system`    | Built-in Interpreter   | This is a simple implementation that understands basic home automation commands like "turn on the light" or "stop the music". It currently supports only English, German, French and Spanish and the vocabulary is still very limited. The exact syntax still needs to be documented, for the moment you need to refer to the [source code](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/text/StandardInterpreter.java#L42). |
+| `system`    | Built-in Interpreter   | This is a simple implementation that understands basic home automation commands like "turn on the light" or "stop the music". Explained in more detail bellow.  |
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:223]: bellow -> below

File: configuration/multimedia.md [hunk 2]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
 listenAndAnswer("javasound", "sonos:PLAY5:kitchen")
 listenAndAnswer("voskstt", "voicerss", "voicerss:frFR_Axel", "system,rulehli", "javasound", "sonos:PLAY5:kitchen", "fr-FR", "listeningItem")
 ```
+
+### The Built-in Interpreter
+
+These interpreter is available by default and includes a build-it grammar for English, German, French and Spanish that you can extend using the voiceSystem metadata. These are some examples of the built-in English grammar:
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:390]: The interpreter is available by default and includes built-in grammar ... that can be extended using ...

Here are some examples ...

File: configuration/multimedia.md [hunk 3]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+play the <item name>
+pause the <item name>
+rewind the <item name>
+fast forward the <item name>
+start the <item name>
+stop the <item name>
+refresh the <item name>
+```
+
+For exact built-in grammar you can refer to the [source code](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/text/StandardInterpreter.java#L42).
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:407]: At least the line 42 is probably not correct. Did you want to point to L92, the beginning of the createRules method?

File: configuration/multimedia.md [hunk 4]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+
+#### Target item
+
+The interpreter resolves the item name based on the its label/synonyms and it's parent label/synonyms.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:412]: The interpreter resolves the Item name based on its label/synonyms and its parent label/synonyms.

File: configuration/multimedia.md [hunk 5]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+An example of the possible situations will be:
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:414]: could be

File: configuration/multimedia.md [hunk 6]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+If you have `Group` item labeled as `TV` with a `Dimmer` child `Brightness`.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:416]: . -> :

File: configuration/multimedia.md [hunk 7]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+The interpreter understands these phrases as the same `turn off tv`, `turn off brightness`, `turn off tv brightness` and `turn off brightness tv`.
+
+If you add a Switch child labeled as `Power` to the group.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:419]: . -> :

File: configuration/multimedia.md [hunk 8]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+The interpreter understands now also the phrases `turn off power`, `turn off tv power` and `turn off power tv`.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:420]: now also understands

File: configuration/multimedia.md [hunk 9]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+But the `turn off tv` phase now detect a collision because two matching items accept the `OFF` command.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:421]: but the `turn off tv` phrase now detects a collision because of two matching Items accepting the `OFF` command.

File: configuration/multimedia.md [hunk 10]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+#### Name prevalence
+
+One way you can solve this is by using the name prevalence, items with start with other items names take prevalence over them.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:425]: In general, please capitalize Item, Thing ...

File: configuration/multimedia.md [hunk 11]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+If the Switch item has the name `tv` and the and Dimmer item the name `tv_brightness` there will be no collisions between them and the `OFF` command will target the Switch item.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:427]: and the Dimmer
and therefore the `OFF` command will

File: configuration/multimedia.md [hunk 12]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+#### Exact match label/synonym prevalence
+
+Other way you can solve this is by using the exact match prevalence, items with have an exact match on their label/synonyms take prevalence.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:431]: Another
items with have an exact  --> sounds not correct?

File: configuration/multimedia.md [hunk 13]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+If the Switch item has the synonym `TV` there will be no collisions between them and the `OFF` command will target the Switch item.
+
+#### Location prevalence
+
+The dialog processor forward its locationItem to the standard interpreter to be used for reducing collisions on the target resolution.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:437]: forwards
location Item

File: configuration/multimedia.md [hunk 14]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+If you have two items labeled as `Light` but one is a child of the location item you have configured for the dialog execution, the item takes prevalence.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:439]: ... but one is a child of the location item that has been configured for ...

File: configuration/multimedia.md [hunk 15]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+### Item custom rules.
+
+You can register custom rules into the interpreter using items and the metadata namespace `voiceSystem`.
+
+The examples of valid rules:
+
+```
+"start? watch|watching $*$ on $name$":
+Matches "start watching some show on tv" sends command "some show".
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:457]: Maybe ->
Matches "start watching some show on tv" and sends command "some show".

File: configuration/multimedia.md [hunk 16]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+"watch|play $*$ on tv":
+Matches "play some show on tv" sends command "some show".
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:460]: see above

File: configuration/multimedia.md [hunk 17]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+"start? watch|watching $cmd$ at|on? $name$":
+Matches "watch some show tv" only if item command description contains `some_show_id=some show`, sends command "some_show_id".
+```
+
+As you can see there are some reserved tokens and characters:
+
+* `$name$` defines the place of the item name (resolved as explained before), is optional.
+* `$cmd$` defines the place of a command label, extracted from the item command description.
+* `$*$` defines the place of a command, whose value is not constrained.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:473]: its value

File: configuration/multimedia.md [hunk 18]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+* `|` defines alternative word tokens.
+* `?` defines optional word token.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:475]: token or tokens?

File: configuration/multimedia.md [hunk 19]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+The item metadata namespace `voiceSystem` allows these configurations:
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:477]: allows following configurations

File: configuration/multimedia.md [hunk 20]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+* **isTemplate**: The rule defined on these item metadata will not target it but similar items (items with same tags and semantic).
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:479]: this item metadata
it -> what does >it< refer to?

File: configuration/multimedia.md [hunk 21]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+* **isSilent**: The interpreter will say nothing in case these rule is executed correctly (util if you plan on trigger a rule and talk from it).
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:480]: in case the rule is executed

>(util if you plan on trigger a rule and talk from it).

Do you mean?
> (until you implement and trigger a rule and talk from within that rule)

File: configuration/multimedia.md [hunk 22]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+* **isForced**: Send the command without check the current Item state (default behavior).
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:481]: without checking

File: configuration/multimedia.md [hunk 23]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+Note that if the `isTemplate` config is false, the rule target is limited to the item that registers it. When it's true the item registering the rule gets excluded of matching with it.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:483]: > When it's true the item registering the rule gets excluded of matching with it.
When it is true

"the rule gets excluded of matching with it" -> I don't understand the meaning, also what is "it" here. Maybe rephrase to be more clear on the meaning.

File: configuration/multimedia.md [hunk 24]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+Note that when you use the option `isTemplate` in rules without the `$name$` token, collisions are still solved based on location. So you can have a `play $cmd$ here` rule which is scoped to the dialog location.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:485]: based on the location.
> So you can have
Therefore, you can have ...

File: configuration/multimedia.md [hunk 25]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+
+There are some limitations:
+
+* Rule should contain `$cmd$` or `$*$` but not both.
+* Rules that include `$name$` and `$*$` should have at least one non optional token between them.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:490]: non-optional

File: configuration/multimedia.md [hunk 26]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+* Rules can not start by `$name$` or `$*$`, neither by them prefixed only of optional tokens.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:491]: Would "must not" be better here?

File: configuration/multimedia.md [hunk 27]
```diff
@@ -386,3 +384,109 @@ listenAndAnswer(null, null)
...
+* Rules can not contain `$name$`, `$cmd$` or `$*$` multiple times.
```
Comment (stefan-hoehn) [inline configuration/multimedia.md:492]: same as above.

**Review by stefan-hoehn — CHANGES_REQUESTED** [review]
Thanks for the comprehensive contribution. See my review comments.

File: configuration/multimedia.md (hunk 20 above)
Comment (GiviMAD) [reply configuration/multimedia.md:479]: It  refers to the item owning the metadata, I have changed it to `itself`.

File: configuration/multimedia.md (hunk 21 above)
Comment (GiviMAD) [reply configuration/multimedia.md:480]: No, what I meant is that the situation the option was added for is when you want to handle the command from a rule and talk from there. 
Because it's weird that the interpreter says "Ok" if you are using the target item to trigger a rule and do extra work there.

An example will be having a command "Search in google $*$" writing to a String Item that you will use to trigger a rule that search programmatically at google and told you something.

Comment (GiviMAD) [reply configuration/multimedia.md:480]: I can change it to :
```
(an possible use case can be trigger a rule on Item command to answer the command programmatically)
```

**Comment by GiviMAD** [comment]
I have applied all the suggestions I have left one open to know if the rephrase is good enough.

For the record all the functionality is already in the 4.1.0 version, I don't know if the documentation is updated on path releases but just in case.

Thank you for the review!

File: configuration/multimedia.md (hunk 21 above)
Comment (stefan-hoehn) [reply configuration/multimedia.md]: "a possible use case can be trigger a rule on Item command to answer programmatically"

-> a possible use case can be a trigger for a rule on an Item command to answer programmatically

**Comment by stefan-hoehn** [comment]
> For the record all the functionality is already in the 4.1.0 version

I have marked as 4.1 then. Thanks for letting me know.

**Comment by GiviMAD** [comment]
> I have marked as 4.1 then. Thanks for letting me know.

You are welcome, I should have it mentioned on the description.

I have applied the latest comment.

**Comment by stefan-hoehn** [comment]
@GiviMAD this is still not fixed

> Another way you can solve this is by using the exact match prevalence, Items whose label/synonym match**es** the one in the command exactly take prevalence.

**Review by stefan-hoehn — APPROVED** [review]
Finally we're done. Thanks for the contribution.

File: configuration/multimedia.md [hunk 28]
```diff
@@ -386,3 +384,105 @@ listenAndAnswer(null, null)
...
+
+#### Name prevalence
+
+One way you can solve this is by using the name prevalence, Items with start with other Items names take prevalence over them.
+
+If the Switch Item has the name `tv` and the Dimmer Item the name `tv_brightness` there will be no collisions between them and therefore the `OFF` command will target the Switch Item.
+
+#### Exact match label/synonym prevalence
+
+Another way you can solve this is by using the exact match prevalence, Items whose label/synonym matches the one in the command exactly take prevalence.
```
[the remark this answers is not in the document]
Comment (GiviMAD) [reply configuration/multimedia.md]: > @GiviMAD Did you see this one?

:/ sorry again for missing it.

And thank you for the review!
